### PR TITLE
Bootstrap4

### PR DESCRIPTION
**Pull request by tordans**
This is a WIP status but all I can do with my limited knowledge of both bootstrap4 and simple form.

If this looks like a good idea, please take over and finish it.

The PR has basically two parts:
- changing the form generator so it works for bootstrap 4 (details inside the extended commit messages)
- adding generator feature so both version 3 and 4 can be used. This part takes ideas from https://github.com/plataformatec/simple_form/issues/1336

### CODE REVIEW

File: lib/generators/simple_form/install_generator.rb [hunk 1]
```diff
@@ -32,8 +35,10 @@ def copy_scaffold_template
       end
 
       def show_readme
-        if behavior == :invoke && options.bootstrap?
-          readme "README"
+        if behavior == :invoke
+          case options
```
Comment (fritzgrabo) [inline lib/generators/simple_form/install_generator.rb]: Whoops, this `case` misses an `end`.

**Comment by SebDE** [comment]
Thanks @tordans for this work - but I have seen there is a problem with the horizontal form layout.
The Bootstrap 4 documentation says:
> Add the .row class to form groups and use the .col-* classes to specify the width of your labels and controls.

So you need to add those class to your form-group divs like

> config.wrappers :horizontal_form, tag: 'div', class: 'form-group row', error_class: 'has-danger' do |b|

or 

> config.wrappers :horizontal_file_input, tag: 'div', class: 'form-group row', error_class: 'has-danger' do |b|

**Comment by tordans** [comment]
@SebDE please feel free to extend the PR. I am aware that there are quite some problems yet but I will only work on those things that I need to fix for my current project – not time for more unfortunately. This should be a collective effort to get it done.

**Comment by kevinelliott** [comment]
I would like to see this work continued!

**Comment by ant0n** [comment]
@kevinelliott you can replace your `config/initialize/simple_form_bootstrap.rb` with [this](https://github.com/plataformatec/simple_form/pull/1337/files) settings.
Anyone know why Travis CI failed? I run tests locally and they are good.

**Comment by secondstreetmedia** [comment]
Just starting playing with this. One small thing I noticed is in the `lib/generators/simple_form/templates/install_generator.rb` file, code line #12 in the `info_bootstrap` method is:

    return if options.bootstrap? || options.foundation?

but should be:

    return if options.bootstrap? || options.bootstrap4? || options.foundation?

Like I said, small, but I thought I'd let you know. 

Thanks for all your efforts.

Cheers!

**Comment by kevin-jj** [comment]
@tordans any updates? look forward to getting it work in the production.

**Comment by tordans** [comment]
@kevin-jj I don't plan to work on this any time soon. As I wrote in my PR-description, someone should take over. Also, the thing is that Bootstrap4 is still alpha and changes a lot. I am pretty sure that this PR is outdated since the last set of changes http://blog.getbootstrap.com/2016/07/27/bootstrap-4-alpha-3/.
Feel free to continue the work.

**Comment by Superpencil** [comment]
With current bootstrap-4 state and updated for current simple-form release

```rb
# Use this setup block to configure all options available in SimpleForm.
SimpleForm.setup do |config|
  config.error_notification_class = 'alert alert-danger'
  config.button_class = 'btn btn-primary'
  config.boolean_label_class = nil

  config.wrappers :vertical_form, tag: 'div', class: 'form-group', error_class: 'has-danger' do |b|
    b.use :html5
    b.use :placeholder
    b.optional :maxlength
    b.optional :minlength
    b.optional :pattern
    b.optional :min_max
    b.optional :readonly
    b.use :label, class: 'form-control-label'

    b.use :input, class: 'form-control'
    b.use :error, wrap_with: { tag: 'span', class: 'form-text text-muted' }
    b.use :hint,  wrap_with: { tag: 'p', class: 'form-text text-muted' }
  end

  config.wrappers :vertical_file_input, tag: 'div', class: 'form-group', error_class: 'has-danger' do |b|
    b.use :html5
    b.use :placeholder
    b.optional :maxlength
    b.optional :minlength
    b.optional :readonly
    b.use :label, class: 'form-control-label'

    b.use :input
    b.use :error, wrap_with: { tag: 'span', class: 'form-text text-muted' }
    b.use :hint,  wrap_with: { tag: 'p', class: 'form-text text-muted' }
  end

  config.wrappers :vertical_boolean, tag: 'div', class: 'form-group', error_class: 'has-danger' do |b|
    b.use :html5
    b.optional :readonly

    b.wrapper tag: 'div', class: 'checkbox' do |ba|
      ba.use :label_input
    end

    b.use :error, wrap_with: { tag: 'span', class: 'form-text text-muted' }
    b.use :hint,  wrap_with: { tag: 'p', class: 'form-text text-muted' }
  end

  config.wrappers :vertical_radio_and_checkboxes, tag: 'div', class: 'form-group', error_class: 'has-danger' do |b|
    b.use :html5
    b.optional :readonly
    b.use :label, class: 'form-control-label'
    b.use :input
    b.use :error, wrap_with: { tag: 'span', class: 'form-text text-muted' }
    b.use :hint,  wrap_with: { tag: 'p', class: 'form-text text-muted' }
  end

  config.wrappers :horizontal_form, tag: 'div', class: 'form-group row', error_class: 'has-danger' do |b|
    b.use :html5
    b.use :placeholder
    b.optional :maxlength
    b.optional :minlength
    b.optional :pattern
    b.optional :min_max
    b.optional :readonly
    b.use :label, class: 'col-md-3 form-control-label'

    b.wrapper tag: 'div', class: 'col-md-9' do |ba|
      ba.use :input, class: 'form-control'
      ba.use :error, wrap_with: { tag: 'span', class: 'form-text text-muted' }
      ba.use :hint,  wrap_with: { tag: 'p', class: 'form-text text-muted' }
    end
  end

  config.wrappers :horizontal_file_input, tag: 'div', class: 'form-group row', error_class: 'has-danger' do |b|
    b.use :html5
    b.use :placeholder
    b.optional :maxlength
    b.optional :minlength
    b.optional :readonly
    b.use :label, class: 'col-md-3 form-control-label'

    b.wrapper tag: 'div', class: 'col-md-9' do |ba|
      ba.use :input
      ba.use :error, wrap_with: { tag: 'span', class: 'form-text text-muted' }
      ba.use :hint,  wrap_with: { tag: 'p', class: 'form-text text-muted' }
    end
  end

  config.wrappers :horizontal_boolean, tag: 'div', class: 'form-group row', error_class: 'has-danger' do |b|
    b.use :html5
    b.optional :readonly

    b.wrapper tag: 'div', class: 'offset-md-3 col-md-9' do |wr|
      wr.wrapper tag: 'div', class: '' do |ba|
        ba.use :label_input
      end

      wr.use :error, wrap_with: { tag: 'span', class: 'form-text text-muted' }
      wr.use :hint,  wrap_with: { tag: 'p', class: 'form-text text-muted' }
    end
  end

  config.wrappers :horizontal_radio_and_checkboxes, tag: 'div', class: 'form-group row', error_class: 'has-danger' do |b|
    b.use :html5
    b.optional :readonly

    b.use :label, class: 'col-md-3 form-control-label'

    b.wrapper tag: 'div', class: 'col-md-9' do |ba|
      ba.use :input
      ba.use :error, wrap_with: { tag: 'span', class: 'form-text text-muted' }
      ba.use :hint,  wrap_with: { tag: 'p', class: 'form-text text-muted' }
    end
  end

  config.wrappers :inline_form, tag: 'div', class: 'form-group', error_class: 'has-danger' do |b|
    b.use :html5
    b.use :placeholder
    b.optional :maxlength
    b.optional :minlength
    b.optional :pattern
    b.optional :min_max
    b.optional :readonly
    b.use :label, class: 'sr-only'

    b.use :input, class: 'form-control'
    b.use :error, wrap_with: { tag: 'span', class: 'form-text text-muted' }
    b.use :hint,  wrap_with: { tag: 'p', class: 'form-text text-muted' }
  end

  config.wrappers :multi_select, tag: 'div', class: 'form-group', error_class: 'has-danger' do |b|
    b.use :html5
    b.optional :readonly
    b.use :label, class: 'form-control-label'
    b.wrapper tag: 'div', class: 'form-inline' do |ba|
      ba.use :input, class: 'form-control'
      ba.use :error, wrap_with: { tag: 'span', class: 'form-text text-muted' }
      ba.use :hint,  wrap_with: { tag: 'p', class: 'form-text text-muted' }
    end
  end
  # Wrappers for forms and inputs using the Bootstrap toolkit.
  # Check the Bootstrap docs (http://getbootstrap.com)
  # to learn about the different styles for forms and inputs,
  # buttons and other elements.
  config.default_wrapper = :vertical_form
  config.wrapper_mappings = {
    check_boxes: :vertical_radio_and_checkboxes,
    radio_buttons: :vertical_radio_and_checkboxes,
    file: :vertical_file_input,
    boolean: :vertical_boolean,
    datetime: :multi_select,
    date: :multi_select,
    time: :multi_select
  }
end
```

**Comment by yordis** [comment]
@tordans a question, how to do `from-control-*` where `*` is the state of the input, for example, `form-control-danger`?

Because I am see that on the v4 spec

**Comment by yordis** [comment]
Nevermind, we need this #1476

**Comment by ryush00** [comment]
Any updates?

**Comment by tordans** [comment]
We startet working on this issue again over at https://github.com/xijo/simple_form-bootstrap4 (@xijo and I). The idea is to have a place to check the generated form elements against the original specs. There is a status list for that project at https://github.com/xijo/simple_form-bootstrap4/issues/1. However, we are not sure if this will be the final approach since handling the simple form config ist quite difficult and there are issues like https://github.com/plataformatec/simple_form/pull/1369 that are in the way.

**Comment by yordis** [comment]
@tordans any updates on this? I checked what you did on that fork but I don't think it works anymore.

- The `error tag` changed
- Normally we should add `was-validated` in the `form` only but this works as well
- `label` changed
- `input` needs a class called `is-invalid` for be able to work

```
config.wrappers :vertical_form, tag: 'div', class: 'form-group', error_class: 'was-validated' do |b|
    b.use :html5
    b.use :placeholder
    b.optional :maxlength
    b.optional :minlength
    b.optional :pattern
    b.optional :min_max
    b.optional :readonly

    b.use :label, class: 'col-form-label'
    b.use :input, class: 'form-control'
    b.use :error, wrap_with: {tag: 'span', class: 'invalid-feedback'}
    b.use :hint, wrap_with: {tag: 'small', class: 'form-text text-muted'}
  end
```

**Comment by tordans** [comment]
@yordis I dont understand your bullets / code example.
At https://v4-alpha.getbootstrap.com/components/forms/#validation there is no talk about 'was-validated' and on https://github.com/xijo/simple_form-bootstrap4/blob/master/config/initializers/simple_form_bootstrap.rb which I referenced above as the current WIP there it not either …

**Comment by yordis** [comment]
@tordans I installed the version from the `npm` and I had to run it locally because it is waaay different from what it's published (Deal with it, Bootstrap life) soo probably you will have to change it again.

I am working on it, so far I had to monkey patched two functions to be able to do it (because it was faster for me to get thing done, definitely bad practice)

**Comment by Atrox** [comment]
@yordis I too installed the `npm` version (`bootstrap@4.0.0-alpha.6`) and it works as expected.
Check your installed bootstrap version, it probably is outdated.

**Comment by yordis** [comment]
@Atrox  Bootstrap 4 Dev Branch
- https://github.com/twbs/bootstrap/blob/v4-dev/scss/_forms.scss#L266
- https://github.com/twbs/bootstrap/blob/v4-dev/scss/_forms.scss#L273
- https://github.com/twbs/bootstrap/blob/v4-dev/scss/mixins/_forms.scss#L28
- https://github.com/twbs/bootstrap/blob/v4-dev/scss/mixins/_forms.scss#L57

`Gemfile: gem 'bootstrap', git: 'https://github.com/twbs/bootstrap-rubygem'`
screen shot 2017-07-16 at 4 37 03 am

I am telling you, it will change everything again, which it doesn't surprise me 😄

**Comment by yordis** [comment]
I am making it work so far with this. Again the monkey patch is just cutting time.

There is some changes from `simple_form` that could actually make more flexible the setup. Like adding more callbacks with the correct contexts so I can send back the correct configuration

```ruby
SimpleForm.setup do |config|
  config.error_notification_class = 'alert alert-danger'
  config.button_class = 'btn btn-primary'
  config.boolean_label_class = nil

  config.wrappers :vertical_form, tag: 'div', class: 'form-group' do |b|
    b.use :html5
    b.use :placeholder
    b.optional :maxlength
    b.optional :minlength
    b.optional :pattern
    b.optional :min_max
    b.optional :readonly

    b.use :label, class: 'col-form-label' do

    end
    b.use :input, class: 'form-control'
    b.use :error, wrap_with: {tag: 'span', class: 'invalid-feedback'}
    b.use :hint, wrap_with: {tag: 'small', class: 'form-text text-muted'}
  end

  config.wrappers :vertical_file_input, tag: 'div', class: 'form-group' do |b|
    b.use :html5
    b.use :placeholder
    b.optional :maxlength
    b.optional :minlength
    b.optional :readonly

    b.use :label, class: 'col-form-label'
    b.use :input
    b.use :error, wrap_with: {tag: 'span', class: 'invalid-feedback'}
    b.use :hint, wrap_with: {tag: 'p', class: 'form-text text-muted'}
  end

  config.wrappers :vertical_boolean, tag: 'div', class: 'form-group' do |b|
    b.use :html5
    b.optional :readonly

    b.wrapper tag: 'div', class: 'checkbox' do |ba|
      ba.use :label_input
    end

    b.use :error, wrap_with: {tag: 'span', class: 'invalid-feedback'}
    b.use :hint, wrap_with: {tag: 'p', class: 'form-text text-muted'}
  end

  config.wrappers :vertical_radio_and_checkboxes, tag: 'div', class: 'form-group' do |b|
    b.use :html5
    b.optional :readonly

    b.use :label, class: 'col-form-label'
    b.use :input
    b.use :error, wrap_with: {tag: 'span', class: 'invalid-feedback'}
    b.use :hint, wrap_with: {tag: 'p', class: 'form-text text-muted'}
  end

  config.wrappers :horizontal_form, tag: 'div', class: 'form-group row' do |b|
    b.use :html5
    b.use :placeholder
    b.optional :maxlength
    b.optional :minlength
    b.optional :pattern
    b.optional :min_max
    b.optional :readonly

    b.use :label, class: 'col-md-3 col-form-label'
    b.wrapper tag: 'div', class: 'col-md-9' do |ba|
      ba.use :input, class: 'form-control'
      ba.use :error, wrap_with: {tag: 'span', class: 'invalid-feedback'}
      ba.use :hint, wrap_with: {tag: 'p', class: 'form-text text-muted'}
    end
  end

  config.wrappers :horizontal_file_input, tag: 'div', class: 'form-group row' do |b|
    b.use :html5
    b.use :placeholder
    b.optional :maxlength
    b.optional :minlength
    b.optional :readonly
    b.use :label, class: 'col-md-3 col-form-label'

    b.wrapper tag: 'div', class: 'col-md-9' do |ba|
      ba.use :input
      ba.use :error, wrap_with: {tag: 'span', class: 'invalid-feedback'}
      ba.use :hint, wrap_with: {tag: 'p', class: 'form-text text-muted'}
    end
  end

  config.wrappers :horizontal_boolean, tag: 'div', class: 'form-group row' do |b|
    b.use :html5
    b.optional :readonly

    b.wrapper tag: 'div', class: 'offset-md-3 col-md-9' do |wr|
      wr.wrapper tag: 'div', class: '' do |ba|
        ba.use :label_input
      end

      wr.use :error, wrap_with: {tag: 'span', class: 'invalid-feedback'}
      wr.use :hint, wrap_with: {tag: 'p', class: 'form-text text-muted'}
    end
  end

  config.wrappers :horizontal_radio_and_checkboxes, tag: 'div', class: 'form-group row' do |b|
    b.use :html5
    b.optional :readonly

    b.use :label, class: 'col-md-3 col-form-label'
    b.wrapper tag: 'div', class: 'col-md-9' do |ba|
      ba.use :input
      ba.use :error, wrap_with: {tag: 'span', class: 'invalid-feedback'}
      ba.use :hint, wrap_with: {tag: 'p', class: 'form-text text-muted'}
    end
  end

  config.wrappers :inline_form, tag: 'div', class: 'form-group' do |b|
    b.use :html5
    b.use :placeholder
    b.optional :maxlength
    b.optional :minlength
    b.optional :pattern
    b.optional :min_max
    b.optional :readonly

    b.use :label, class: 'sr-only'
    b.use :input, class: 'form-control'
    b.use :error, wrap_with: {tag: 'span', class: 'invalid-feedback'}
    b.use :hint, wrap_with: {tag: 'p', class: 'form-text text-muted'}
  end

  config.wrappers :multi_select, tag: 'div', class: 'form-group' do |b|
    b.use :html5
    b.optional :readonly

    b.use :label, class: 'col-form-label'
    b.wrapper tag: 'div', class: 'form-inline' do |ba|
      ba.use :input, class: 'form-control'
      ba.use :error, wrap_with: {tag: 'span', class: 'invalid-feedback'}
      ba.use :hint, wrap_with: {tag: 'p', class: 'form-text text-muted'}
    end
  end


  config.default_wrapper = :vertical_form
  config.wrapper_mappings = {
      check_boxes: :vertical_radio_and_checkboxes,
      radio_buttons: :vertical_radio_and_checkboxes,
      file: :vertical_file_input,
      boolean: :vertical_boolean,
      datetime: :multi_select,
      date: :multi_select,
      time: :multi_select
  }
end

# SimpleForm::ActionViewExtensions::FormHelper.class_eval do
#
#   private
#
#   def simple_form_css_class(record, options)
#     html_options = options[:html]
#     as = options[:as]
#
#     if html_options.key?(:class)
#       html_options[:class]
#     elsif record.is_a?(String) || record.is_a?(Symbol)
#       as || record
#     else
#       form_css_class_from_record(record, as)
#     end
#   end
#
#   def form_css_class_from_record(record, as)
#     classes = []
#
#     record = record.last if record.is_a?(Array)
#     action = record.respond_to?(:persisted?) && record.persisted? ? :edit : :new
#     classes << as ? "#{action}_#{as}" : dom_class(record, action)
#
#     classes << 'was-validated' if record.errors.any?
#
#     classes.join(' ')
#   end
# end

SimpleForm::Inputs::Base.class_eval do
  # We just need the ability to pass a block on the `use` function
  # then we save it on the Leaf class to be able to call it
  def merge_wrapper_options(options, wrapper_options)
    if wrapper_options
      wrapper_options.merge(options) do |key, oldval, newval|
        case key.to_s
          when 'class'
            classes = Array(oldval) + Array(newval)
            classes << 'is-invalid' if has_errors?
            classes
          when 'data', 'aria'
            oldval.merge(newval)
          else
            newval
        end
      end
    else
      options
    end
  end
end
```

**Comment by mendelk** [comment]
[Bootstrap 4 Beta](http://blog.getbootstrap.com/2017/08/10/bootstrap-4-beta/)

**Comment by 4ndv** [comment]
Any changes needed for the code above to work with bs4 beta?

**Comment by mrfoto** [comment]
Yes - Bootstrap 4 requires error_class to be present on `input` so #1476 needs to be merged first. And there are a couple of class renamings that happened. `col-sm-offset-3` -> `ml-auto` for example

**Comment by Superpencil** [comment]
1. @plataformatec doesn't use simple_form anymore
2. It has been stated simple_form will never add support for specific frameworks again

I'd say the chances of this ever being resolved are very slim.

Our best hope is @rafaelfranca allowing someone else to submit a proper PR for bootstrap-4 support.

Both my and @yordis's solutions are half-assed attempts at basic support. Bootstrap also has many input-group items and custom controls that we need to write brand new wrappers for.

I'm willing to do this work if we can get some sort of signal from a Member that it would be considered for merge?

**Comment by rainerborene** [comment]
> @plataformatec doesn't use simple_form anymore

@Superpencil why? and what's the alternative?

**Comment by Superpencil** [comment]
@rainerborene not using `ruby` but `elixir`, `JavaScript` or `Go` it seems

**Comment by tordans** [comment]
@feliperenan I saw you where doing some issue-gardening yesterday. Could you join this discussion about the future of simple_form for bootstrap 4? There are a few things that need to change in the simple_form core. Would you be willing to review and merge those changes (more config options are required for example).

**Comment by feliperenan** [comment]
Hi @tordans 

I'll take a closer look as soon as possible.

> There are a few things that need to change in the simple_form core.

Do you know or could you list the things necessary?

**Comment by tordans** [comment]
@feliperenan it's been a long while … but from the top of my head:

1. the changes of this PR are outdated but in general still required in a very similar way.
2. bootstrap4 requires a handling of classes like error classes that is not supported by simple form ATM. An example is the monkey path in https://github.com/plataformatec/simple_form/pull/1337#issuecomment-315594533. 
3. Another example is https://github.com/xijo/simple_form-bootstrap4/commit/c7bbcde697d2ed80c270d703d71ee8c59929c9a8#diff-b674edbbd1a9d89a9ae3d301ed2ff506L48. There seems to be a solution for one of the issues in simple form but it is only implemented for some html elements (select but not input or something like this?); simple form probably needs more global configs flags to add classes and wrappers.
4. Another example is https://github.com/plataformatec/simple_form/pull/1369#issuecomment-312208316 which might be solved the way described in this comment or by some global config variable.
4. In general there are quite few cases where the current simple_form wrapper-features are not enough to generate a html-code that is valid for bootstrap 4 (last time I checked bootstrap was still beta)

The bigger question is IMO: 
Are you guys still actively working on this gem and are you open to change it so bootstrap 4 can be implemented in a sane way?
Will you invest time to do this yourselves?
Or start the community process around it?
I ask this questions b/c of comments like https://github.com/plataformatec/simple_form/pull/1337#issuecomment-328436704 and the fact that it has been extremely silent on this issue until now.

**Comment by feliperenan** [comment]
> Are you guys still actively working on this gem and are you open to change it so bootstrap 4 can be implemented in a sane way?

Lately we have looked for bugs and security issues. However we are accepting features proposal normally.

> Will you invest time to do this yourselves? Or start the community process around it?

I don't have intention to implement it. If the community thinks it is a good idea we can discuss about and approve if that's the case.

**My opinion about bootstrap 4:**

I'm not sure if it's a good idea create an intializer to bootstrap 4 since it's beta yet and I don't know how much it will change until a stable version is released.

These days a lot of CSS frameworks are becoming more popular among developers, and the community has been asking us to support it. We believe simple_form should support custom components - [see issue](https://github.com/plataformatec/simple_form/issues/1176). That means the community can add support for different frameworks in other gems.

c/c @tegon @rafaelfranca

**Comment by rafaelfranca** [comment]
I have the same opinion as @feliperenan. When I was working in the bootstrap 3 support I regret a lot of adding first class support to bootstrap built-in in simple form. Today I'd do differently and have a simple_form-bootstrap gem maintained by the community but make simple form flexible enough that it was possible to build those kind of integrations.

I'm happy to accept changes to make simple form flexible in order to support bootstrap 4, but I have no interest on building this support myself or even maintaining it.

**Comment by feliperenan** [comment]
Closing it since we won't add support to Bootstrap 4.

We are working on [simple_form-boostrap](https://github.com/rafaelfranca/simple_form-bootstrap/pull/56) to update to Bootstrap 4 and making Simple Form flexible to support it. Feel free to joins to us.

Thanks

**Comment by tordans** [comment]
If anyone finds this thread: My follow up questions is at https://github.com/plataformatec/simple_form/issues/1336#issuecomment-375728201.

**Comment by feliperenan** [comment]
Hey @tordans, we will update the Bootstrap initializer to use the last version of Bootstrap, you can see more about this decision here https://github.com/plataformatec/simple_form/issues/1561.

Since you were doing the work here, are you willing to finish this PR? I think Simple Form has all support for Bootstrap 4 given the last PRs.

**Comment by feliperenan** [comment]
Just to make it clear. We will just update the existing Bootstrap initializer to use the version 4.

**Comment by feliperenan** [comment]
@m5o Would you like continue this work?

**Comment by m5o** [comment]
@feliperenan Will open a new PR shortly, targeting `v4.1` 👍

**Comment by yordis** [comment]
@feliperenan what is the status on this topic?

I noticed the project that you are working on but it is a full Rails project so it is hard to extract the pieces that actually make simple_form flexible from it.

**Comment by feliperenan** [comment]
@yordis The last version of Simple Form has all support to Bootstrap 4. You can use that another project as guide.

**Pull request closed without merging**